pinot/generative/torch_gvae/decoder.py

```diff
@@ -0,0 +1,119 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+class InnerProductDecoder(nn.Module):
+    """Decoder for using inner product for edge prediction."""
+
+    def __init__(self, dropout):
+        super(InnerProductDecoder, self).__init__()
+        self.dropout = dropout
+
+    def forward(self, z):
+        """ Returns a symmetric adjacency matrix of size (N,N)
+        where A[i,j] = probability there is an edge between nodes i and j
+        """
+        z = F.dropout(z, self.dropout, training=self.training)
+        adj = torch.mm(z, z.t())
+        return adj
+
+
+class EdgeAndNodeDecoder(nn.Module):
+    """ Decoder that returns both a predicted adjacency matrix
+        and node identities
+    """
+    def __init__(self, feature_dim, num_atom_types, hidden_dim=64, dropout=0):
+        super(EdgeAndNodeDecoder, self).__init__()
+        self.dropout = dropout
+        self.hidden_dim = hidden_dim
+        self.decode_nodes = nn.Sequential(
+            nn.Linear(feature_dim, self.hidden_dim),
+            nn.ReLU(),
+            nn.Linear(self.hidden_dim, num_atom_types)
+        )
+
+    def forward(self, z):
+        """
+        Arg:
+            z (FloatTensor):
+                Shape (N, hidden_dim)
+        Returns:
+            (adj, node_preds)
+                adj (FloatTensor): has shape (N, N) is a matrix where entry (i.j)
+                    stores the probability that there is an edge between atoms i,j
+                node_preds (FloatTensor): has shape (N, num_atom_types) where each
+                    row i stores the probability of the identity of atom i
+        """
+        z_prime = F.dropout(z, self.dropout, training=self.training)
+        adj = torch.mm(z_prime, z_prime.t())
+        node_preds = self.decode_nodes(z_prime)
+        return (adj, node_preds)
+
+
+class SequentialDecoder(nn.Module):
+    """ Simple decoder where the node identity prediction is dependent
+    on the adjacency matrix.
+    Attributes
+    ----------
+    z_to_zx : torch.nn.Module
+        neural networks that map general latent space encoding $z$
+        to that for node prediction $z_x$
+    z_to_za : torch.nn.Module
+        neural networks that map general latent space encoding $z$
+        to that for edge prediction $z_a$
+    zx_to_x : torch.nn.Module
+        neural networks that map latent encoding of node $z_x$ to node
+        predictions $\hat{x}$
+    """
+
+    def __init__(self, embedding_dim, num_atom_types):
+        super(SequentialDecoder, self).__init__()
+        self.embedding_dim = embedding_dim
+        self.num_atom_types = num_atom_types
+
+        self.Dx1 = 64
```

Comment:
let's not hard code anything here. I think it's good enough to leave `z_to_zx` and friends as modules.

or, we can have them to be either torch models or a list of config string in the style of `pinot.representation`

Reply:
yup, will move these into constructor variables